### ISSUE TITLE
Any experience to help me optimize on a particular problem size(a small M)?

### ISSUE THREAD

**Issue by HuaYZhao** [NVIDIA/cutlass#1309]
I am a beginner to cutlass and I have reviewed many related documents and examples；I also have a general understanding of ThreadBlockShape, WarpShape, InstructionShape, NumStages;
In my current problem domain, I want to deal with matrix multiplication with M, N, K of [8,4096,12288]; And M may vary between 2 and 8;In the flash decoding++ paper, this is called the flat gemm problem;
With the above background, I have the following questions:
1、I have used cutlass profiler to search for the best parameters, but sometimes I manually change ThreadBlockShape, WarpShape, InstructionShape, resulting in compilation errors. I want to know what the constraints are on setting these parameters, and if there is documentation or a tutorial on how to tune them; Or is the combination of these parameters already mandatory in the definition of the header file? 
2、The size of M in ThreadBlockShape is usually greater than or equal to 32. If my problem size M is 8, will the kernel fill with 0 during calculation? This seems to cause a certain waste. I am not sure how much this cost is. Is it possible and necessary to set M in ThreadBlockShape to 8 to get the best performance for this size problem, and if so, what should I do to make this adjustment?

In other words, I currently seem to understand the concept of these parameters, but it is not clear how these parameters affect the specific computation during kernel operation, including the allocation and utilization of resources. Is there any relevant documentation or tutorial available to learn, appreciate not only!

**Comment by hwu36** [comment]
i assume you are using fp16 on a100.  if so, the instruction shape is always 16x8x16.  you problem size M is super small.  as you said, resources would be waste.  you may have to use some irregular tile sizes that are not listed in the profiler.  these irregular ones may or may not work.  you need to try them.

you need to set threadblock M as small as possible, but bigger than your problem size M.  16 is minimum since it cannot be smaller than the instruction shape M, but I am not sure if it works in cutlass now.  32 is more likely working.  thread block K is 32 or 64.  warp block K is the same as thread block K.  Normally, we use 4 warps per thread block.  warp number = threadblock_M / warp_M x threadblock_N / warp_N.  you can try to set warp M as the same as threadblock M or half of threadblock M and then choose warp N and threadblock N accordingly.

you have big problem size K.  so to get good perf, you need to use splitk or streamk.

**Comment by HuaYZhao** [comment]
Thank you for your recovery, it helped me dispel part of the fog, I will try again; I do work on fp16 now;

>  assume you are using fp16 on a100. if so, the instruction shape is always 16x8x16.

Does that mean there's no way to change it? Or is it strongly not recommended to change it? In addition, how does this parameter work on other cards such as A10 or 3090, is there documentation available, or is it bound to SM?

> In my current problem domain, I want to deal with matrix multiplication with M, N, K of [8,4096,12288];

I made a description error, I actually have a large N, the actual size of the problem should be M, N, K=[8, 12288, 4096]；

> you need to set threadblock M as small as possible, but bigger than your problem size M. 16 is minimum since it cannot be smaller than the instruction shape M, but I am not sure if it works in cutlass now. 32 is more likely working. thread block K is 32 or 64. warp block K is the same as thread block K. Normally, we use 4 warps per thread block. warp number = threadblock_M / warp_M x threadblock_N / warp_N. you can try to set warp M as the same as threadblock M or half of threadblock M and then choose warp N and threadblock N accordingly.

This is where I'm confused, I think you have a richer understanding to give this advice, is it purely empirical understanding, or are there some rules? In other words, I would like to have a primer on how to set these parameters so that when the size of my problem changes, say to [8, 4096, 4096] or [8, 27648, 5120], etc., I can make my own design based on some experience!

It's better to teach a man fishing than to give him fish.
Thank you very much again!

**Comment by hwu36** [comment]
> Does that mean there's no way to change it? Or is it strongly not recommended to change it? In addition, how does this parameter work on other cards such as A10 or 3090, is there documentation available, or is it bound to SM?

all sm80 should use this for fp16.

> I made a description error, I actually have a large N, the actual size of the problem should be M, N, K=[8, 12288, 4096]；

what i said earlier still holds.  you have small M and large K.

> This is where I'm confused, I think you have a richer understanding to give this advice, is it purely empirical understanding, or are there some rules? In other words, I would like to have a primer on how to set these parameters so that when the size of my problem changes, say to [8, 4096, 4096] or [8, 27648, 5120], etc., I can make my own design based on some experience!

What i described can apply to all small M and large K problems.

**Comment by HuaYZhao** [comment]
I used stream_k mode for my calculations and found that executing `device_gemm()` was actually relatively faster, but the total time taken by these lines of code in this mode was not negligible; In my case, the calculation time was roughly 6.5088e-02 ms, but the workspace related time was 4.50528e-01 ms,This is especially true in stream k mode

```
    size_t workspace_size = DeviceKernel::get_workspace_size(arguments);
    cutlass::device_memory::allocation<uint8_t> workspace(workspace_size);

    DeviceKernel device_gemm;
    cutlass::Status status = device_gemm.initialize(arguments,
                                                workspace.get());
```

I am currently wrapping a matrix multiplication operator, adjusting the layout for my matrix size, and finally exposing it to python as pybind; Whether to open the workspace every time matrix multiplication is called seems to cause a lot of performance consumption, and the overall time is very long to observe on the python side. Are there any optimizations?

I got good performance under the example 47_ampere_gemm_universal_streamk, but couldn't encapsulate it as an interface while retaining its high performance because of this workspace!

**Comment by hwu36** [comment]
don't put workspace allocation in the critical path.  you can manage your memory yourself or run this part in parallel with something else.
